lib/logger.js

```diff
@@ -0,0 +1,653 @@
+
+    // Defaults for instantiation parameters
+    this.level = 'INFO'
+    this.url = constants.LOGDNA_URL
+    this.flushLimit = constants.FLUSH_BYTE_LIMIT
+    this.flushIntervalMs = constants.FLUSH_INTERVAL_MS
+    this.shimProperties = undefined
+    this.indexMeta = false
+    this.app = 'default'
+    this.env = undefined
+    this.baseBackoffMs = constants.BASE_BACKOFF_MILLIS
+    this.maxBackoffMs = constants.MAX_BACKOFF_MILLIS
+
+    // Initialize internal instance variables
+    this[kLineLengthTotal] = 0
+    this[kBuffer] = []
+    this[kMeta] = {}
+    this[kIsLoggingBackedOff] = false
+    this[kAttempts] = 0
+    this[kRequestDefaults] = undefined
+    this[kFlusher] = null
+    this[kReadyToSend] = []
+    this[kIsSending] = false
+    this[kTotalLinesReady] = 0
+    this[kBackoffMs] = constants.BASE_BACKOFF_MILLIS
+
+    let useHttps = true
+    let withCredentials = false
+    let tags = undefined
+    let timeout = constants.DEFAULT_REQUEST_TIMEOUT
+    let hostname = os.hostname()
+    let mac = undefined
+    let ip = undefined
+
+    if (has(options, 'level')) {
+      const level = options.level.toUpperCase()
+      if (!constants.LOG_LEVELS.includes(level)) {
+        const err = new Error('Invalid level')
+        err.meta = {
+          got: level
+        , expectedOneOf: constants.LOG_LEVELS
+        }
+        throw err
+      }
+      this.level = level
+    }
+
+    if (has(options, 'tags')) {
+      tags = options.tags
+      if (typeof tags === 'string') {
+        tags = tags.split(constants.TAGS_RE)
+      }
+      if (!Array.isArray(tags)) {
+        const err = new TypeError('tags should be passed as a String or an Array')
+        err.meta = {got: options.tags}
+        throw err
+      }
+      tags = tags
+        .filter((tag) => { return tag !== '' })
+        .map((tag) => { return tag.trim() })
+        .join(',')
+    }
+
+    if (has(options, 'meta')) {
+      const meta = options.meta
+      if (typeOf(meta) !== 'object') {
+        const err = new TypeError('meta needs to be an object of key-value pairs')
+        err.meta = {got: meta}
+        throw err
+      }
+      this[kMeta] = {...meta}
+    }
+
+    if (has(options, 'timeout')) {
+      if (!Number.isInteger(options.timeout)) {
+        const err = new TypeError('timeout must be an Integer')
+        err.meta = {got: options.timeout}
+        throw err
+      }
+      if (options.timeout > constants.MAX_REQUEST_TIMEOUT) {
+        throw new Error(
+          `timeout cannot be longer than ${constants.MAX_REQUEST_TIMEOUT}`
+        )
+      }
+      timeout = options.timeout
+    }
+
+    if (has(options, 'hostname')) {
+      checkStringParam(options.hostname, 'Hostname')
+      if (!constants.HOSTNAME_CHECK.test(options.hostname)) {
+        throw new Error('Invalid hostname')
+      }
+      hostname = options.hostname
+    }
+
+    if (has(options, 'mac')) {
+      checkStringParam(options.mac, 'MAC Address')
```

Comment:
I think we can also be more specific so the mac addresses used are actually mac addresses

Reply:
this is validated below